phac-nml/irida-next
# Namespace to group controller service

## What does this PR do and why?
_Describe in detail what your merge request does and why._

This pull request adds in functionality (controller and service) to allow sharing of a namespace with a group and to unshare a previously shared namespace with a group. This will allow us to share namespaces with other groups with a set max role rather than adding in individual users from other groups if we want to allow access to the namespace.

**Note: Currently the controller actions for share and unshare do a redirect but will be updated to use turbo stream in the UI branch**

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

## How to set up and validate locally
_Numbered steps to set up and validate the change are strongly suggested._

1. Open up rails console from the terminal with `bin/rails c`
2. Find a bunch of group and project ids that you will use to link groups to groups and projects with groups
3. Try and create some `NamespaceGroupLink` objects using the `Namespaces::GroupShareService` linking groups with other groups and projects with groups using the ids from step 2. Make sure to try out duplicate entries, entries where the group id doesn't exist in the `namespaces` table
4. Try unsharing the links from step 3 using the `Namespaces::GroupUnhareService`. Make sure to also try out unsharing links that don't exist in the table
5. Ensure the controller tests are working correctly for sharing/unsharing by reviewing the tests

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.

  the [PR acceptance checklist](https://phac-nml.github.io/irida-next/docs/development/development_processes/code_review#acceptance-checklist) for this PR.
